package/boot/uboot-envtools/files/qualcommax_ipq50xx

```
@@ -8,6 +8,11 @@ touch /etc/config/ubootenv
 board=$(board_name)
 
 case "$board" in
+glinet,gl-b3000)
+	idx="$(find_mtd_index 0:APPSBLENV)"
+	[ -n "$idx" ] && \
+		ubootenv_add_uci_config "/dev/mtd$idx" "0x0" "0x40000" "0x20000" "2"
```

No need to specify number of blocks (last parameter). It's auto detected.

---

@testuser7  @georgemoussalem 

Could it be that a uboot_env entry for this unit is not necessary ?

in the stock firmware, the entry is found in /etc/fw_env.conifg, witch contains the following
```
/dev/mtd6 0x0 0x40000 0x20000 2
```
from the bootlog this would suggest the "0:APPSBLENV" partition which i used for the uboot_env entry as shown above
```
[    1.108066] 0x000000000000-0x000000080000 : "0:SBL1"                   0
[    1.114863] 0x000000080000-0x000000100000 : "0:MIBIB"                 1
[    1.119879] 0x000000100000-0x000000140000 : "0:BOOTCONFIG"   2
[    1.124580] 0x000000140000-0x000000240000 : "0:QSEE"                   3
[    1.131000] 0x000000240000-0x000000280000 : "0:DEVCFG"              4
[    1.135227] 0x000000280000-0x0000002c0000 : "0:CDT"                      5
[    1.139838] 0x0000002c0000-0x000000340000 : "0:APPSBLENV"        6
[    1.145267] 0x000000340000-0x000000480000 : "0:APPSBL"
[    1.151682] 0x000000480000-0x000000580000 : "0:ART"
[    1.157095] 0x000000580000-0x000000600000 : "0:TRAINING"
[    1.160404] 0x000000600000-0x000000800000 : "CFG"
[    1.168282] 0x000000800000-0x000008000000 : "rootfs"
```

but with or without the last parameter, the following error is seen during boot

[    6.039129] DSA: tree 0 setup
Cannot parse config file '/etc/fw_env.config': No such file or directory
Failed to find NVMEM device
[    6.742041] nss-dp 39d00000.dp2 eth1: PHY Link up speed: 1000
[    6.744777] qca8k 90000.mdio-1:11 lan1: configuring for phy/gmii link mode
[    6.747157] adpt_mp_port_netdev_change_notify[1186]:ERROR:netdev change notify with incorrect port 0
[    6.753591] ssdk_dev_event[2313]:ERROR:netdev change notify failed
```
if I manualy add the file to ~/target/linux/qualcommax/ipq50xx/base-files/etc/fw_env.config 

the error is gone . 

Should i remove the entry from unoot_env ? if so, should i add as above ?

---

@georgemoussalem 
 
any thoughts on this ?

---

if the fw_env.config file is in fact necessary, perhaps I can add it via a uci-defaults script instead ?

---

this happens when loading the image in ram, not when loading from nand

---

I have also tried hard coding the uboot_env entry like so ...
```
glinet,gl-b3000)
	ubootenv_add_uci_config "/dev/mtd6" "0x0" "0x40000" "0x20000"
	;;
```

same results

```
[    6.000204] DSA: tree 0 setup
Cannot parse config file '/etc/fw_env.config': No such file or directory
Failed to find NVMEM device
```

the only thing thats silences the error is manually adding the file, as done it my last commit.

---

ok, i have tried every combination, the onmly thing that satisfies it is adding the file manually. Currently being adding in 
```
/target/linux/qualcommax/ipq50xx/base-files/etc/fw_env.config
```

The strange part is, without manually adding the file as above, even though it complains with the error, the file is created anyway ...

```
root@OpenWrt:~# ls /etc/
TZ                    hotplug.json          protocols
apk                   init.d                rc.button
banner                inittab               rc.common
banner.failsafe       iproute2              rc.d
board.d               localtime             rc.local
board.json            luci-uploads          resolv.conf
capabilities          modules-boot.d        services
config                modules.conf          shadow
crontabs              modules.d             shells
device_info           mtab                  shinit
diag.sh               nftables.d            ssl
dnsmasq.conf          odhcp6c.user          sysctl.conf
dropbear              odhcp6c.user.d        sysctl.d
e2fsck.conf           openwrt_release       syslog.conf
ethers                openwrt_version       sysupgrade.conf
fstab                 os-release            uci-defaults
fw_env.config         passwd                udhcpc.user
group                 ppp                   udhcpc.user.d
hosts                 preinit               urandom.seed
hotplug-preinit.json  profile
hotplug.d             profile.d
root@OpenWrt:~# cat /etc/fw_env.config
/dev/mtd6 0x0 0x40000 0x20000
```

so the file is there, but a reboot and the error is still seen 

```
[    6.149883] DSA: tree 0 setup
Cannot parse config file '/etc/fw_env.config': No such file or directory
Failed to find NVMEM device
```

perhaps this error is toothless ?  in which case manually adding the file is unnecessary ?

From what i can tell, it does not effect performance with or without the error.  So I guess the question that remains is ... to add or not to add,  the file.